hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java

```diff
@@ -807,16 +809,17 @@ private ClosableIteratorWithSchema<HoodieRecord> getRecordsIterator(
 
     Option<Pair<Function<HoodieRecord, HoodieRecord>, Schema>> schemaEvolutionTransformerOpt =
         composeEvolvedSchemaTransformer(dataBlock);
+
     // In case when schema has been evolved original persisted records will have to be
     // transformed to adhere to the new schema
-    if (schemaEvolutionTransformerOpt.isPresent()) {
-      return ClosableIteratorWithSchema.newInstance(
-          new CloseableMappingIterator<>(blockRecordsIterator,
-              schemaEvolutionTransformerOpt.get().getLeft()),
-          schemaEvolutionTransformerOpt.get().getRight());
-    } else {
-      return ClosableIteratorWithSchema.newInstance(blockRecordsIterator, dataBlock.getSchema());
-    }
+    Function<HoodieRecord, HoodieRecord> transformer =
+        schemaEvolutionTransformerOpt.map(Pair::getLeft)
+            .orElse(Function.identity());
+
+    Schema schema = schemaEvolutionTransformerOpt.map(Pair::getRight)
+        .orElse(dataBlock.getSchema());
+
```

Comment:
The code looks clean but for use case when `schemaEvolutionTransformerOpt.isPresent()` is false, there is unnecessary invocation of `Function.identity()`.

Reply:
Not sure i understand what your concern is: we now duplicate much less code at the expense of having empty transformer (which would be optimized out by the JVM)